### DISCUSSION TITLE
Autoscaling colormap

### DISCUSSION THREAD

**Discussion by hrbigelow** [bokeh/bokeh#13090] [Internals and design (Q&A)]
On Bryan's suggestion I wanted to start a discussion here related to this [post](https://discourse.bokeh.org/t/auto-scaling-colormap/10363)

With streaming data, x and y axes generally automatically adjust to a changing data range.  This is not the case for a colormap however.  Could it be?  Conceptually this would be identical - compute the min and max after an update, then update the `low` and `high` values of the colormap object accordingly.  In particular I am thinking of a server-side mechanism that works with no interaction from the user.

As I understand it, this might be tricky.  Any thoughts would be appreciated!

Thanks,

Henry

**Comment by bryevdv** [comment]
cc @bokeh/dev for any thoughts. Currently only `image` does anything like this, and (AFAIK) it also still only does it on a per-image basis (no global colormapping across all images). I think an auto-scaling colormap is a completely reasonable ask,  my main concerns are how to do it without introducing too much internal coupling. E.g. I wouldn't really want colorbars to hold direct references to data sources, if that can at all be avoided.

**Reply by ianthomas23** [reply]
There is a related use case for the recently added `image_stack` glyph. When we add support for selecting a single 2D image at a time from the 3D image stack, the should be the option of scaling the colormap by the 2D image limits or the overall limits of the whole 3D stack. But this isn't much of a jump from the current `image` colormapping, we'll just have the option of scaling by the `image[:, :, :]` or `image[:, :, i]` limits.
